src/lib/updating-element.ts

```diff
@@ -0,0 +1,501 @@
+    if (attrValue !== undefined) {
+      const attr = ctor.attributeNameForProperty(name);
+      if (attr !== undefined) {
+        // track the attr name/value being set to be able to avoid
+        // reflecting back to the property setter via attributeChangedCallback.
+        this._serializingInfo = [attr, attrValue];
+        if (attrValue === null) {
+          this.removeAttribute(attr);
+        } else {
+          this.setAttribute(attr, attrValue);
+        }
+        this._serializingInfo = null;
+      }
+    }
+  }
+
+  private _attributeToProperty(name: string, value: string) {
+    // Use tracking info to avoid deserializing attribute value if it was
+    // just set from a property setter.
+    if (!this._serializingInfo ||
+        (this._serializingInfo[0] !== name && this._serializingInfo[1] !== value)) {
+      const ctor = (this.constructor as typeof UpdatingElement);
+      const propName = ctor._attributeToPropertyMap[name];
+      this[propName as keyof this] = ctor.propertyValueFromAttribute(propName, value);
+    }
+  }
+
+  /**
+   * Invalidates the element causing it to asynchronously update regardless
+   * of whether or not any property changes are pending. This method is
+   * automatically called when any registered property changes. Returns a Promise
+   * that resolves when the element has finished updating.
+   */
+  async invalidate() {
+    // Do not re-queue validation if already invalid (pending) or currently updating.
+    if (this._isPendingUpdate()) {
+      return this._validatePromise;
+    }
+    this._validationState = ValidationState.Invalid;
+    // Make a new validate promise and put the resolver on the stack.
+    this._validatePromise = new Promise((resolve) => this._validateResolvers.push(resolve));
+    // Wait a tick to actually process changes (allows batching).
+    await 0;
+    // Mixin instance properties once, if they exist.
+    if (this._instanceProperties) {
+      Object.assign(this, this._instanceProperties);
+      this._instanceProperties = null;
+    }
+    // Rip off changedProps.
+    const changedProps = this._changedProps || {};
+    this._changedProps = null;
+    if (this.shouldUpdate(changedProps)) {
+      // During update, setting properties does not trigger invalidation.
+      this._validationState = ValidationState.Updating;
+      this.update(changedProps);
+      this._validationState = ValidationState.Valid;
+      // During finishUpdate, setting properties does trigger invalidation,
+      // and users may choose to await other state, like children being updated.
+      await this.finishUpdate(changedProps);
+    } else {
+      this._validationState = ValidationState.Valid;
+    }
+    // Only resolve the promise if we finish in a valid state (finishUpdate
+    // did not trigger more work).
+    if (this._validationState === ValidationState.Valid) {
+      while (this._validateResolvers.length) {
```

Comment:
Since you resolve all the Promises at once, is seems like we could just share a single Promise. Are they ever resolved separately?

Reply:
Yeah, that had occurred to me as well and I forgot to clean it up. Will do that.

Reply:
Done.